Custom uniform search optimization

Teivaz opened the pull request:
- Switched ProgramState from unordered_map(string, Uniform) to unordered_map(GLint, Uniform) for faster uniform search. This will give opportunity to access uniforms in program state by their location that gives faster search speed than using strings.

Teivaz commented:
I double checked everything and don't see any reasons for fialing build on ios and mac. According to log buildmachine failed to locate "v3-deps-2.zip" that is not affected by changes in pull request.

minggo commented:
Thanks.
It is not your fault.

dabingnn commented:
@Teivaz 
This patch sounds good. 
Have you tried that how much performance gain could we get if we use this patch for rendering?

Teivaz commented:
Here what I mesured with CC_PROFILER in uniform set up part of the GAFPlayer library (flash animations playback in cocos2d-x)

Win32 Debug:

```
GAFSpriteWithAlpha::setUniforms ::  
avg1: 0µ,  avg2: 29µ, min: 0µ,   max: 2000µ,    total: 3.57s,   nr calls: 119298
```

Win32 Debug optimized:

```
GAFSpriteWithAlpha::setUniforms ::  
avg1: 3µ,  avg2: 6µ,  min: 0µ,   max: 1000µ,    total: 0.74s,   nr calls: 119298
```

Win32 Release:

```
GAFSpriteWithAlpha::setUniforms ::  
avg1: 0µ,  avg2: 1µ,  min: 0µ,   max: 1000µ,    total: 0.19s,   nr calls: 119298
```

Win32 Release optimized:

```
GAFSpriteWithAlpha::setUniforms ::  
avg1: 0µ,  avg2: 0µ,  min: 0µ,   max: 1000µ,    total: 0.04s,   nr calls: 119298
```

dabingnn commented:
Great, this is a good improvement on WIN32 platform.
We appreciate if you have performance results on mobile platform.
Thanks.

Teivaz commented:
Debug ipod touch 4:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 29µ, avg2: 30µ, min: 25µ,  max: 4143µ,    total: 1.42s,   nr calls: 46410
```

Debug ipod touch 4 optimized:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 8µ,  avg2: 9µ,  min: 7µ,   max: 1282µ,    total: 0.44s,   nr calls: 46410
```

Release ipod touch 4:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 17µ, avg2: 17µ, min: 14µ,  max: 4148µ,    total: 0.82s,   nr calls: 46410
```

Release ipod touch 4 optimized:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 5µ,  avg2: 5µ,  min: 4µ,   max: 1274µ,    total: 0.25s,   nr calls: 46410
```

Teivaz commented:
Debug ipad mini:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 18µ, avg2: 20µ, min: 17µ,  max: 2098µ,    total: 0.95s,   nr calls: 46410
```

Debug ipad mini optimized:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 5µ,  avg2: 6µ,  min: 5µ,   max: 553µ, total: 0.31s,   nr calls: 46410
```

Release ipad mini:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 10µ, avg2: 12µ, min: 9µ,   max: 800µ, total: 0.58s,   nr calls: 46410
```

Release ipad mini optimized:

```
cocos2d: GAFSpriteWithAlpha::setUniforms :: 
avg1: 3µ,  avg2: 3µ,  min: 2µ,   max: 1176µ,    total: 0.17s,   nr calls: 46410
```

dabingnn commented:
Cool, thanks.

Teivaz commented:
Updated due to new changes in current branch. 
@dabingnn can this request get in the v3.2 please?

dabingnn commented on cocos/renderer/CCGLProgramState.cpp:
```diff
@@ -329,9 +330,9 @@ void GLProgramState::apply(const Mat4& modelView)
     if(_uniformAttributeValueDirty)
     {
-        for(auto& uniformValue : _uniforms)
+        for(auto& uniformValue : _uniformsByName)
```
`uniformValue` is not a value now, it is better use a new name.

dabingnn commented:
@Teivaz 
Thanks, I added some comments.
I think it could be merge safely.
And far more, it is better add a test case for this new feature.
It would be cool if you can do it, or you can tell me, I would add some test cast for this.

Teivaz committed:
Updated pull request #7065 due to new changes in current branch
- Switched ProgramState from unordered_map<string, Uniform> to unordered_map<GLint, Uniform> for faster uniform search
- changed `uniformValue` variable name to more relative

Teivaz commented:
@dabingnn, I updated request so now that varible name is more relative to what it is.
I used this functionality in my code in the following way:

In initialization of shaders uniform location is being stored:

```
GLint m_colorTransformMultLocation;
m_colorTransformMultLocation = program->getUniformLocationForName("colorTransformMult");
```

then before object's render task is being queued the uniforms are set up in the following way:

```
getGLProgramState()->setUniformVec4(m_colorTransformMultLocation, m_colorTransformMult);
```

instead of

```
getGLProgramState()->setUniformVec4("colorTransformMult", m_colorTransformMult);
```

Meanwhile the old function is also valid but as I mentioned before is slower than new one.
